// @taiga-ui/addon-commerce/esm2022/pipes/decimal/decimal.pipe.mjs
import { inject, INJECTOR, Injector, Pipe } from '@angular/core';
import { TuiAmountPipe } from '@taiga-ui/addon-commerce/pipes/amount';
import { TUI_NUMBER_FORMAT } from '@taiga-ui/core/tokens';
import { map, switchMap } from 'rxjs';
import * as i0 from "@angular/core";
class TuiDecimalPipe {
    constructor() {
        this.format = inject(TUI_NUMBER_FORMAT);
        this.amountPipe = Injector.create({
            providers: [{ provide: TuiAmountPipe }],
            parent: inject(INJECTOR),
        }).get(TuiAmountPipe);
    }
    transform(value, currency = '') {
        return this.format.pipe(switchMap((format) => this.amountPipe.transform(value, currency).pipe(map((value) => {
            const [, decimal] = value.split(format.decimalSeparator);
            return decimal ? `${format.decimalSeparator}${decimal}` : '';
        }))));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDecimalPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiDecimalPipe, isStandalone: true, name: "tuiDecimal" }); }
}
export { TuiDecimalPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDecimalPipe, decorators: [{
            type: Pipe,
            args: [{ standalone: true, name: 'tuiDecimal' }]
        }] });
