// @taiga-ui/core/esm2022/components/alert/alert.tokens.mjs
import { inject, InjectionToken } from '@angular/core';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { TUI_NOTIFICATION_OPTIONS } from '@taiga-ui/core/components/notification';
import { BehaviorSubject, combineLatest, map, of } from 'rxjs';
export const TUI_ALERT_DEFAULT_OPTIONS = {
    autoClose: 3000,
    label: '',
    closeable: true,
    data: undefined,
};
export const TUI_ALERT_OPTIONS = new InjectionToken(ngDevMode ? 'TUI_ALERT_OPTIONS' : '', {
    factory: () => ({
        ...TUI_ALERT_DEFAULT_OPTIONS,
        ...inject(TUI_NOTIFICATION_OPTIONS),
    }),
});
export const TUI_ALERT_POSITION = new InjectionToken(ngDevMode ? 'TUI_ALERT_POSITION' : '', {
    factory: () => (inject(TUI_IS_MOBILE) ? '1rem 1rem 0 auto' : '2rem 3rem 0 auto'),
});
export const TUI_ALERTS = new InjectionToken(ngDevMode ? 'TUI_ALERTS' : '', {
    factory: () => new BehaviorSubject([]),
});
/**
 * Grouping alerts by their component
 */
export const TUI_ALERTS_GROUPED = new InjectionToken(ngDevMode ? 'TUI_ALERTS_GROUPED' : '', {
    factory: () => combineLatest([
        of(new Map()),
        inject(TUI_ALERTS),
    ]).pipe(map(([map, alerts]) => {
        map.forEach((_, key) => map.set(key, []));
        alerts.forEach((alert) => {
            const key = alert.component.component;
            const value = map.get(key) || [];
            map.set(key, [...value, alert]);
        });
        return Array.from(map.values());
    })),
});
export function tuiAlertOptionsProvider(options) {
    return {
        provide: TUI_ALERT_OPTIONS,
        useFactory: () => ({
            ...TUI_ALERT_DEFAULT_OPTIONS,
            ...(inject(TUI_ALERT_OPTIONS, { optional: true, skipSelf: true }) ||
                inject(TUI_NOTIFICATION_OPTIONS)),
            ...options,
        }),
    };
}
