// @taiga-ui/kit/esm2022/components/multi-select/multi-select-option/multi-select-option.component.mjs
import { NgIf } from '@angular/common';
import { ChangeDetectionStrategy, Component, computed, inject } from '@angular/core';
import { tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiOptionWithValue } from '@taiga-ui/core/components/data-list';
import { TUI_TEXTFIELD_OPTIONS } from '@taiga-ui/core/components/textfield';
import { TUI_ITEMS_HANDLERS, } from '@taiga-ui/core/directives/items-handlers';
import { TuiCheckbox } from '@taiga-ui/kit/components/checkbox';
import { tuiInjectValue } from '@taiga-ui/kit/utils';
import * as i0 from "@angular/core";
class TuiMultiSelectOption {
    constructor() {
        this.option = inject(TuiOptionWithValue, {
            optional: true,
        });
        this.handlers = inject(TUI_ITEMS_HANDLERS);
        this.value = tuiInjectValue();
        this.size = inject(TUI_TEXTFIELD_OPTIONS).size;
        this.selected = computed((controlValue = this.value(), optionValue = this.option?.value()) => tuiIsPresent(optionValue) &&
            tuiIsPresent(controlValue) &&
            controlValue.some((item) => this.handlers?.identityMatcher()(item, optionValue) ??
                item === optionValue));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOption, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiMultiSelectOption, isStandalone: true, selector: "ng-component", ngImport: i0, template: `
        <input
            *ngIf="option"
            tuiCheckbox
            type="checkbox"
            [checked]="selected()"
            [size]="size() === 'l' ? 'm' : 's'"
        />
    `, isInline: true, styles: [":host-context(tui-dropdown-mobile) input{order:100;margin-inline-start:auto}\n"], dependencies: [{ kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: TuiCheckbox, selector: "input[type=\"checkbox\"][tuiCheckbox]", inputs: ["size"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiMultiSelectOption };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOption, decorators: [{
            type: Component,
            args: [{ standalone: true, imports: [NgIf, TuiCheckbox], template: `
        <input
            *ngIf="option"
            tuiCheckbox
            type="checkbox"
            [checked]="selected()"
            [size]="size() === 'l' ? 'm' : 's'"
        />
    `, changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host-context(tui-dropdown-mobile) input{order:100;margin-inline-start:auto}\n"] }]
        }] });
