// @taiga-ui/legacy/esm2022/components/input-time/input-time.options.mjs
import { InjectionToken } from '@angular/core';
import { tuiProvideOptions } from '@taiga-ui/cdk/utils/miscellaneous';
/**
 * TODO(v5): delete it
 * @deprecated use new version of {@link https://taiga-ui.dev/components/input-time#options TuiInputTime} (from @taiga-ui/kit) instead
 */
export const MAX_TIME_VALUES = {
    HH: 23,
    MM: 59,
    SS: 59,
    MS: 999,
};
/**
 * TODO(v5): delete it
 * @deprecated use new version of {@link https://taiga-ui.dev/components/input-time#options TuiInputTime} (from @taiga-ui/kit) instead
 */
export const TUI_INPUT_TIME_DEFAULT_OPTIONS = {
    icon: () => '@tui.clock',
    mode: 'HH:MM',
    maxValues: MAX_TIME_VALUES,
    itemSize: 'm',
    nativePicker: false,
};
/**
 * TODO(v5): delete it
 * Default parameters for InputTime component
 * @deprecated use new version of {@link https://taiga-ui.dev/components/input-time#options TuiInputTime} (from @taiga-ui/kit) instead
 */
export const TUI_INPUT_TIME_OPTIONS = new InjectionToken(ngDevMode ? 'TUI_INPUT_TIME_OPTIONS' : '', {
    factory: () => TUI_INPUT_TIME_DEFAULT_OPTIONS,
});
/**
 * TODO(v5): delete it
 * @deprecated use new version of {@link https://taiga-ui.dev/components/input-time#options TuiInputTime} (from @taiga-ui/kit) instead
 */
export function tuiInputTimeOptionsProvider(options) {
    return tuiProvideOptions(TUI_INPUT_TIME_OPTIONS, options, TUI_INPUT_TIME_DEFAULT_OPTIONS);
}
