// @taiga-ui/legacy/esm2022/components/input-time/input-time.directive.mjs
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost } from '@taiga-ui/legacy/classes';
import { tuiAsTextfieldHost } from '@taiga-ui/legacy/tokens';
import * as i0 from "@angular/core";
/**
 * TODO(v5): delete it
 * @deprecated use new version of {@link https://taiga-ui.dev/components/input-time TuiInputTime} (from @taiga-ui/kit) instead
 */
class TuiInputTimeDirective extends AbstractTuiTextfieldHost {
    get value() {
        return this.host.computedValue;
    }
    get mode() {
        return this.host.mode;
    }
    get items() {
        return this.host.items;
    }
    onValueChange(value) {
        if (!value) {
            this.host.nativeValue = '';
        }
        this.host.onValueChange(value);
    }
    process(input) {
        input.inputMode = this.host.inputMode;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTimeDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputTimeDirective, selector: "tui-input-time", providers: [tuiAsTextfieldHost(TuiInputTimeDirective)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiInputTimeDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTimeDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'tui-input-time',
                    providers: [tuiAsTextfieldHost(TuiInputTimeDirective)],
                }]
        }] });
