// @taiga-ui/kit/esm2022/components/tabs/tabs-horizontal.directive.mjs
import { __decorate } from "tslib";
import { ContentChildren, Directive, forwardRef, inject, Input, } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { MutationObserverService, WA_MUTATION_OBSERVER_INIT, } from '@ng-web-apis/mutation-observer';
import { EMPTY_QUERY } from '@taiga-ui/cdk/constants';
import { tuiZonefree } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiPure, tuiPx } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiTab } from './tab.directive';
import { TuiTabsDirective } from './tabs.directive';
import { TUI_TABS_OPTIONS } from './tabs.options';
import * as i0 from "@angular/core";
import * as i1 from "./tabs.directive";
class TuiTabsHorizontal {
    constructor() {
        this.el = tuiInjectElement();
        this.options = inject(TUI_TABS_OPTIONS);
        this.tabs = inject(TuiTabsDirective);
        this.children = EMPTY_QUERY;
        this.sub = inject(MutationObserverService, { self: true })
            .pipe(tuiZonefree(), takeUntilDestroyed())
            .subscribe(() => this.refresh());
        this.underline = this.options.underline;
    }
    ngAfterViewChecked() {
        this.scrollTo(this.tabs.activeItemIndex);
        this.refresh();
    }
    onKeyDownArrow(current, step) {
        this.tabs.moveFocus(current, step);
    }
    refresh() {
        const { activeElement } = this.tabs;
        if (activeElement && !activeElement.isConnected) {
            return;
        }
        const { offsetLeft = 0, offsetWidth = 0 } = activeElement || {};
        this.el.style.setProperty('--t-left', tuiPx(offsetLeft));
        this.el.style.setProperty('--t-width', tuiPx(offsetWidth));
    }
    scrollTo(index) {
        const element = this.tabs.tabs[index];
        if (!element) {
            return;
        }
        const { offsetLeft, offsetWidth } = element;
        if (offsetLeft < this.el.scrollLeft) {
            this.el.scrollLeft = offsetLeft;
        }
        if (offsetLeft + offsetWidth > this.el.scrollLeft + this.el.offsetWidth) {
            this.el.scrollLeft = offsetLeft + offsetWidth - this.el.offsetWidth;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTabsHorizontal, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTabsHorizontal, isStandalone: true, selector: "tui-tabs:not([vertical]), nav[tuiTabs]:not([vertical])", inputs: { underline: "underline" }, host: { listeners: { "animationend": "refresh()", "keydown.arrowRight.prevent": "onKeyDownArrow($event.target, 1)", "keydown.arrowLeft.prevent": "onKeyDownArrow($event.target, -1)" }, properties: { "class._underline": "underline", "style.--t-color": "underline === true ? 'var(--tui-background-accent-1)' : underline" } }, providers: [
            MutationObserverService,
            {
                provide: WA_MUTATION_OBSERVER_INIT,
                useValue: {
                    childList: true,
                    characterData: true,
                    subtree: true,
                },
            },
        ], queries: [{ propertyName: "children", predicate: i0.forwardRef(function () { return TuiTab; }) }], hostDirectives: [{ directive: i1.TuiTabsDirective, inputs: ["activeItemIndex", "activeItemIndex", "size", "size"], outputs: ["activeItemIndexChange", "activeItemIndexChange"] }], ngImport: i0 }); }
}
__decorate([
    tuiPure
], TuiTabsHorizontal.prototype, "scrollTo", null);
export { TuiTabsHorizontal };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTabsHorizontal, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tui-tabs:not([vertical]), nav[tuiTabs]:not([vertical])',
                    providers: [
                        MutationObserverService,
                        {
                            provide: WA_MUTATION_OBSERVER_INIT,
                            useValue: {
                                childList: true,
                                characterData: true,
                                subtree: true,
                            },
                        },
                    ],
                    hostDirectives: [
                        {
                            directive: TuiTabsDirective,
                            inputs: ['activeItemIndex', 'size'],
                            outputs: ['activeItemIndexChange'],
                        },
                    ],
                    host: {
                        '[class._underline]': 'underline',
                        '[style.--t-color]': "underline === true ? 'var(--tui-background-accent-1)' : underline",
                        '(animationend)': 'refresh()',
                        '(keydown.arrowRight.prevent)': 'onKeyDownArrow($event.target, 1)',
                        '(keydown.arrowLeft.prevent)': 'onKeyDownArrow($event.target, -1)',
                    },
                }]
        }], propDecorators: { children: [{
                type: ContentChildren,
                args: [forwardRef(() => TuiTab)]
            }], underline: [{
                type: Input
            }], scrollTo: [] } });
