// @taiga-ui/addon-commerce/fesm2022/taiga-ui-addon-commerce-tokens.mjs
import { InjectionToken } from '@angular/core';
import { tuiExtractI18n } from '@taiga-ui/i18n/utils';
import { of } from 'rxjs';

/**
 * Number and card number i18n
 */
const TUI_CARD_NUMBER_TEXTS = new InjectionToken(ngDevMode ? 'TUI_CARD_NUMBER_TEXTS' : '', {
    factory: tuiExtractI18n('cardNumber'),
});
/**
 * Expiry and card expiry i18n
 */
const TUI_CARD_EXPIRY_TEXTS = new InjectionToken(ngDevMode ? 'TUI_CARD_EXPIRY_TEXTS' : '', {
    factory: tuiExtractI18n('cardExpiry'),
});
/**
 * Card CVC number text [mobile, desktop]
 */
const TUI_CARD_CVC_TEXTS = new InjectionToken(ngDevMode ? 'TUI_CARD_CVC_TEXTS' : '', {
    factory: () => of(['CVC', 'CVC/CVV']),
});

const TUI_PAYMENT_SYSTEM_ICONS = new InjectionToken(ngDevMode ? 'TUI_PAYMENT_SYSTEM_ICONS' : '', {
    factory: () => ({
        mir: '@tui.mir',
        visa: '@tui.visa',
        electron: '@tui.electron',
        mastercard: '@tui.mastercard',
        maestro: '@tui.maestro',
        amex: '@tui.amex',
        dinersclub: '@tui.diners-club',
        discover: '@tui.discover',
        humo: '@tui.humo',
        jcb: '@tui.jcb',
        rupay: '@tui.ru-pay',
        unionpay: '@tui.union-pay',
        uzcard: '@tui.uzcard',
        verve: '@tui.verve',
    }),
});

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_CARD_CVC_TEXTS, TUI_CARD_EXPIRY_TEXTS, TUI_CARD_NUMBER_TEXTS, TUI_PAYMENT_SYSTEM_ICONS };
