// @taiga-ui/core/esm2022/directives/items-handlers/items-handlers.validator.mjs
import { Directive, effect, inject } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { TUI_ALLOW_SIGNAL_WRITES } from '@taiga-ui/cdk/constants';
import { TuiValidator } from '@taiga-ui/cdk/directives/validator';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiItemsHandlersDirective } from './items-handlers.directive';
import * as i0 from "@angular/core";
class TuiItemsHandlersValidator extends TuiValidator {
    constructor() {
        super(...arguments);
        this.handlers = inject(TuiItemsHandlersDirective);
        this.update = effect(() => {
            this.handlers.disabledItemHandler();
            this.onChange();
        }, TUI_ALLOW_SIGNAL_WRITES);
        this.disabledItemHandler = (value) => Array.isArray(value)
            ? value.some((item) => this.handlers.disabledItemHandler()(item))
            : Boolean(value) && this.handlers.disabledItemHandler()(value);
        this.validate = ({ value }) => this.disabledItemHandler(value) ? { tuiDisabledItem: value } : null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiItemsHandlersValidator, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiItemsHandlersValidator, isStandalone: true, providers: [tuiProvide(NG_VALIDATORS, TuiItemsHandlersValidator, true)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiItemsHandlersValidator };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiItemsHandlersValidator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    providers: [tuiProvide(NG_VALIDATORS, TuiItemsHandlersValidator, true)],
                }]
        }] });
