// @taiga-ui/kit/esm2022/components/input-chip/input-chip.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { tuiAsControl, TuiControl } from '@taiga-ui/cdk/classes';
import { TuiActiveZone } from '@taiga-ui/cdk/directives/active-zone';
import { TuiNativeValidator } from '@taiga-ui/cdk/directives/native-validator';
import { TUI_IS_MOBILE, tuiFallbackValueProvider } from '@taiga-ui/cdk/tokens';
import { tuiGetClipboardDataText, tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiDirectiveBinding } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiAsTextfieldAccessor, TuiTextfieldBase, TuiTextfieldMultiComponent, } from '@taiga-ui/core/components/textfield';
import { TuiDropdownDirective, TuiDropdownOpen, tuiDropdownOpen, } from '@taiga-ui/core/directives/dropdown';
import { TUI_ITEMS_HANDLERS, } from '@taiga-ui/core/directives/items-handlers';
import { filter } from 'rxjs';
import { TUI_INPUT_CHIP_OPTIONS } from './input-chip.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/directives/native-validator";
import * as i2 from "@taiga-ui/core/components/textfield";
// TODO(v5): remove base component after angular update
class TuiInputChipBaseDirective extends TuiControl {
    constructor() {
        super(...arguments);
        this.options = inject(TUI_INPUT_CHIP_OPTIONS);
        this.mobile = inject(TUI_IS_MOBILE);
        this.dropdown = inject(TuiDropdownDirective);
        this.textfield = inject(TuiTextfieldMultiComponent);
        this.open = tuiDropdownOpen();
        this.handlers = inject(TUI_ITEMS_HANDLERS);
        this.enabled = tuiDirectiveBinding(TuiDropdownOpen, 'tuiDropdownEnabled', this.interactive, {});
        this.sub = inject(TuiActiveZone)
            .tuiActiveZoneChange.pipe(filter((active) => !active && !this.el.matches('select')), takeUntilDestroyed())
            .subscribe(() => {
            this.onEnter();
            this.textfield.value.set('');
        });
        this.separator = this.options.separator;
        this.unique = this.options.unique;
        this.el = tuiInjectElement();
    }
    setValue(value) {
        this.textfield.value.set('');
        this.onChange(this.unique ? Array.from(new Set(value.reverse())).reverse() : value);
    }
    onEnter() {
        const value = this.textfield.value().trim();
        const items = this.separator ? value.split(this.separator) : [value];
        const valid = items.filter((item) => item && !this.handlers.disabledItemHandler()(item));
        if (!value || !valid.length) {
            return;
        }
        this.setValue([...this.value(), ...valid]);
        this.scrollTo();
    }
    onInput() {
        this.open.set(!!this.dropdown.content);
        if (this.separator && this.textfield.value().match(this.separator)) {
            this.onEnter();
        }
        else {
            this.scrollTo();
        }
    }
    onPaste(event) {
        const value = 'dataTransfer' in event
            ? event.dataTransfer?.getData('text/plain') || ''
            : tuiGetClipboardDataText(event);
        this.textfield.value.set(value);
        this.onEnter();
    }
    onBackspace(key) {
        // (keydown.backspace) doesn't emit event on empty input in ios safari
        if (key === 'Backspace' && !this.textfield.value() && this.interactive()) {
            if (this.mobile || !this.textfield.item) {
                this.onChange(this.value().slice(0, -1));
            }
            else {
                this.el.dispatchEvent(new KeyboardEvent('keydown', {
                    key: 'ArrowLeft',
                    bubbles: true,
                }));
            }
        }
    }
    scrollTo() {
        const sign = this.textfield.el.matches('[dir="rtl"] :scope') ? -1 : 1;
        // Allow change detection to run and add new tag to DOM
        setTimeout(() => {
            this.textfield.el.scrollTo({
                left: sign * Number.MAX_SAFE_INTEGER,
                top: Number.MAX_SAFE_INTEGER,
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputChipBaseDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputChipBaseDirective, isStandalone: true, inputs: { separator: "separator", unique: "unique" }, host: { attributes: { "enterkeyhint": "enter" }, listeners: { "keydown.enter.prevent": "onEnter()", "keydown.zoneless": "onBackspace($event.key)", "input": "onInput()", "paste.prevent": "onPaste($event)", "drop.prevent": "onPaste($event)" }, properties: { "disabled": "disabled()" } }, usesInheritance: true, ngImport: i0 }); }
}
export { TuiInputChipBaseDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputChipBaseDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    host: {
                        enterkeyhint: 'enter',
                        '[disabled]': 'disabled()',
                        '(keydown.enter.prevent)': 'onEnter()',
                        '(keydown.zoneless)': 'onBackspace($event.key)',
                        '(input)': 'onInput()',
                        '(paste.prevent)': 'onPaste($event)',
                        '(drop.prevent)': 'onPaste($event)',
                    },
                }]
        }], propDecorators: { separator: [{
                type: Input
            }], unique: [{
                type: Input
            }] } });
class TuiInputChipDirective extends TuiInputChipBaseDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputChipDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputChipDirective, isStandalone: true, selector: "input[tuiInputChip]", providers: [
            tuiAsControl(TuiInputChipDirective),
            tuiFallbackValueProvider([]),
            tuiAsTextfieldAccessor(TuiInputChipDirective),
        ], usesInheritance: true, hostDirectives: [{ directive: i1.TuiNativeValidator }, { directive: i2.TuiTextfieldBase, inputs: ["invalid", "invalid", "focused", "focused", "readOnly", "readOnly", "state", "state"] }], ngImport: i0 }); }
}
export { TuiInputChipDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputChipDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'input[tuiInputChip]',
                    providers: [
                        tuiAsControl(TuiInputChipDirective),
                        tuiFallbackValueProvider([]),
                        tuiAsTextfieldAccessor(TuiInputChipDirective),
                    ],
                    hostDirectives: [
                        TuiNativeValidator,
                        {
                            directive: TuiTextfieldBase,
                            inputs: ['invalid', 'focused', 'readOnly', 'state'],
                        },
                    ],
                }]
        }] });
