// @taiga-ui/addon-commerce/esm2022/components/input-card-group/input-card-group.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { TuiBreakpointService } from '@taiga-ui/core/services';
import { BehaviorSubject, combineLatest, map } from 'rxjs';
import * as i0 from "@angular/core";
class TuiInputCardGroupDirective {
    constructor() {
        this.c$ = new BehaviorSubject(false);
        this.m$ = inject(TuiBreakpointService).pipe(map((b) => b === 'mobile'));
        this.compact$ = combineLatest([this.c$, this.m$]).pipe(map((c) => c.some(Boolean)));
        this.compact = toSignal(this.compact$, { initialValue: false });
    }
    set compactSetter(compact) {
        this.c$.next(compact);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputCardGroupDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputCardGroupDirective, isStandalone: true, inputs: { compactSetter: ["compact", "compactSetter"] }, host: { properties: { "class._compact": "compact()" } }, ngImport: i0 }); }
}
export { TuiInputCardGroupDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputCardGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    host: { '[class._compact]': 'compact()' },
                }]
        }], propDecorators: { compactSetter: [{
                type: Input,
                args: ['compact']
            }] } });
