// @taiga-ui/experimental/fesm2022/taiga-ui-experimental-components-expand.mjs
import { NgIf, NgTemplateOutlet } from '@angular/common';
import * as i0 from '@angular/core';
import { inject, ChangeDetectorRef, signal, TemplateRef, Component, ChangeDetectionStrategy, ContentChild, Input } from '@angular/core';
import { TuiItem } from '@taiga-ui/cdk/directives/item';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';

class TuiExpand {
    constructor() {
        this.el = tuiInjectElement();
        this.cdr = inject(ChangeDetectorRef);
        this.signal = signal(false);
        this.animating = signal(false);
    }
    set expanded(expanded) {
        if (expanded === this.signal()) {
            return;
        }
        this.signal.set(expanded);
        // TODO: try removing in Angular 17
        this.cdr.detectChanges();
        this.el.classList.toggle('_expanded', expanded);
    }
    onTransitionEnd({ propertyName }) {
        if (propertyName === 'grid-template-rows') {
            this.animating.set(this.signal());
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiExpand, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiExpand, isStandalone: true, selector: "tui-expand", inputs: { expanded: "expanded" }, host: { listeners: { "transitionend.self": "onTransitionEnd($event)" }, properties: { "class._expanded": "signal()" } }, queries: [{ propertyName: "content", first: true, predicate: TuiItem, descendants: true, read: TemplateRef }], ngImport: i0, template: `
        <div class="t-wrapper">
            <ng-container
                *ngIf="signal() || animating()"
                [ngTemplateOutlet]="content || null"
            />
            <ng-content />
        </div>
    `, isInline: true, styles: [":host{transition-property:grid-template-rows,padding;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:relative;display:grid;grid-template-rows:0fr;transition-delay:1ms}:host:not(._expanded){padding-block:0}:host._expanded{grid-template-rows:1fr}:host._expanded .t-wrapper{opacity:1;visibility:visible}.t-wrapper{transition-property:all;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;grid-row:1 / span 2;overflow:hidden;opacity:0;visibility:hidden}\n"], dependencies: [{ kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiExpand, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'tui-expand', imports: [NgIf, NgTemplateOutlet], template: `
        <div class="t-wrapper">
            <ng-container
                *ngIf="signal() || animating()"
                [ngTemplateOutlet]="content || null"
            />
            <ng-content />
        </div>
    `, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class._expanded]': 'signal()',
                        '(transitionend.self)': 'onTransitionEnd($event)',
                    }, styles: [":host{transition-property:grid-template-rows,padding;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:relative;display:grid;grid-template-rows:0fr;transition-delay:1ms}:host:not(._expanded){padding-block:0}:host._expanded{grid-template-rows:1fr}:host._expanded .t-wrapper{opacity:1;visibility:visible}.t-wrapper{transition-property:all;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;grid-row:1 / span 2;overflow:hidden;opacity:0;visibility:hidden}\n"] }]
        }], propDecorators: { content: [{
                type: ContentChild,
                args: [TuiItem, { read: TemplateRef }]
            }], expanded: [{
                type: Input
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiExpand };
