// @taiga-ui/experimental/fesm2022/taiga-ui-experimental-components-accordion.mjs
import { TuiItem } from '@taiga-ui/cdk/directives/item';
import { TuiExpand } from '@taiga-ui/experimental/components/expand';
import * as i0 from '@angular/core';
import { inject, EventEmitter, signal, Directive, Input, Output, DestroyRef, Component, ViewEncapsulation, ChangeDetectionStrategy, ContentChildren } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { EMPTY_QUERY } from '@taiga-ui/cdk/constants';
import * as i1$1 from '@taiga-ui/core/directives/group';
import { tuiGroupOptionsProvider, TuiGroup } from '@taiga-ui/core/directives/group';
import { ReplaySubject, merge, combineLatest, withLatestFrom, map } from 'rxjs';
import { tuiDirectiveBinding } from '@taiga-ui/cdk/utils';
import * as i1 from '@taiga-ui/core/components/button';
import { TuiButton, tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import { tuiAvatarOptionsProvider } from '@taiga-ui/kit/components/avatar';
import * as i2 from '@taiga-ui/kit/directives/chevron';
import { TuiChevron } from '@taiga-ui/kit/directives/chevron';

class TuiAccordionDirective {
    constructor() {
        this.accordion = inject(TuiAccordionComponent);
        this.size = tuiDirectiveBinding(TuiButton, 'size', this.accordion.size);
        this.tuiAccordion = '';
        this.tuiAccordionChange = new EventEmitter();
        this.open = tuiDirectiveBinding(TuiChevron, 'tuiChevron', signal(false));
    }
    ngOnChanges() {
        this.open.set(!!this.tuiAccordion);
        this.accordion.toggle(this);
    }
    toggle() {
        this.open.set(!this.open());
        this.tuiAccordion = this.open();
        this.tuiAccordionChange.emit(this.open());
        this.accordion.toggle(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiAccordionDirective, isStandalone: true, selector: "button[tuiAccordion]", inputs: { tuiAccordion: "tuiAccordion" }, outputs: { tuiAccordionChange: "tuiAccordionChange" }, host: { attributes: { "tuiButton": "", "tuiAccordion": "", "type": "button" }, listeners: { "click": "toggle()" }, properties: { "class._open": "open()" } }, providers: [
            tuiAvatarOptionsProvider({ size: 's' }),
            tuiButtonOptionsProvider({ appearance: 'outline-grayscale' }),
        ], usesOnChanges: true, hostDirectives: [{ directive: i1.TuiButton }, { directive: i2.TuiChevron }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'button[tuiAccordion]',
                    providers: [
                        tuiAvatarOptionsProvider({ size: 's' }),
                        tuiButtonOptionsProvider({ appearance: 'outline-grayscale' }),
                    ],
                    hostDirectives: [TuiButton, TuiChevron],
                    host: {
                        tuiButton: '',
                        tuiAccordion: '',
                        type: 'button',
                        '[class._open]': 'open()',
                        '(click)': 'toggle()',
                    },
                }]
        }], propDecorators: { tuiAccordion: [{
                type: Input
            }], tuiAccordionChange: [{
                type: Output
            }] } });

class TuiAccordionComponent {
    constructor() {
        this.destroyRef = inject(DestroyRef);
        this.toggle$ = new ReplaySubject(Infinity);
        this.expands = EMPTY_QUERY;
        this.directives = EMPTY_QUERY;
        this.closeOthers = true;
        this.size = signal('l');
    }
    set sizeSetter(size) {
        this.size.set(size);
    }
    ngAfterViewInit() {
        merge(combineLatest([this.directives.changes, this.expands.changes]).pipe(withLatestFrom(this.toggle$), map(([, accordion]) => accordion)), this.toggle$)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe((accordion) => this.expand(accordion));
    }
    toggle(directive) {
        this.toggle$.next(directive);
    }
    expand(accordion) {
        if (this.closeOthers && accordion.open()) {
            this.expands.forEach((expand) => {
                expand.expanded = false;
            });
            this.directives.forEach((dir) => {
                if (dir === accordion) {
                    return;
                }
                dir.open.set(false);
                dir.tuiAccordion = false;
                dir.tuiAccordionChange.emit(false);
            });
        }
        const expand = this.expands.get(this.directives.toArray().indexOf(accordion));
        if (expand) {
            expand.expanded = accordion.open();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiAccordionComponent, isStandalone: true, selector: "tui-accordion", inputs: { closeOthers: "closeOthers", sizeSetter: ["size", "sizeSetter"] }, host: { properties: { "attr.data-size": "size()" } }, providers: [
            tuiGroupOptionsProvider({
                orientation: 'vertical',
                collapsed: true,
            }),
        ], queries: [{ propertyName: "expands", predicate: TuiExpand }, { propertyName: "directives", predicate: TuiAccordionDirective }], hostDirectives: [{ directive: i1$1.TuiGroup }], ngImport: i0, template: '<ng-content />', isInline: true, styles: ["tui-accordion{inline-size:-webkit-fill-available;inline-size:-moz-available;inline-size:stretch;border-radius:var(--t-group-radius);font:var(--tui-font-text-m);overflow:hidden}tui-accordion>[tuiAccordion]{text-align:start;justify-content:flex-start}tui-accordion>[tuiAccordion]:last-of-type:not(._open){-webkit-mask-image:var(--t-group-mask-end);mask-image:var(--t-group-mask-end);border-bottom-left-radius:inherit;border-bottom-right-radius:inherit}tui-accordion>[tuiAccordion]:first-child{border-top-left-radius:inherit!important;border-top-right-radius:inherit!important}tui-accordion>[tuiAccordion]:first-child:last-of-type:not(._open){-webkit-mask:none;mask:none}tui-accordion>[tuiAccordion]:after{margin-inline-start:auto!important}tui-accordion>[tuiAccordion]._open+tui-expand{visibility:visible}tui-accordion>[tuiAccordion]>tui-avatar,tui-accordion>[tuiAccordion]>[tuiAvatar]{margin:0}tui-accordion>tui-expand{visibility:hidden;min-block-size:1px!important;padding:1.25rem;transition-property:grid-template-rows,padding,visibility!important;box-shadow:inset 0 0 0 1px var(--tui-border-normal)}tui-accordion>tui-expand:last-child{border-bottom-left-radius:inherit!important;border-bottom-right-radius:inherit!important}tui-accordion[data-size=m]{font:var(--tui-font-text-s)}tui-accordion[data-size=m]>tui-expand{padding:1rem}tui-accordion[data-size=s]{font:var(--tui-font-text-s)}tui-accordion[data-size=s]>tui-expand{padding:.625rem}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'tui-accordion', template: '<ng-content />', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, providers: [
                        tuiGroupOptionsProvider({
                            orientation: 'vertical',
                            collapsed: true,
                        }),
                    ], hostDirectives: [TuiGroup], host: {
                        '[attr.data-size]': 'size()',
                    }, styles: ["tui-accordion{inline-size:-webkit-fill-available;inline-size:-moz-available;inline-size:stretch;border-radius:var(--t-group-radius);font:var(--tui-font-text-m);overflow:hidden}tui-accordion>[tuiAccordion]{text-align:start;justify-content:flex-start}tui-accordion>[tuiAccordion]:last-of-type:not(._open){-webkit-mask-image:var(--t-group-mask-end);mask-image:var(--t-group-mask-end);border-bottom-left-radius:inherit;border-bottom-right-radius:inherit}tui-accordion>[tuiAccordion]:first-child{border-top-left-radius:inherit!important;border-top-right-radius:inherit!important}tui-accordion>[tuiAccordion]:first-child:last-of-type:not(._open){-webkit-mask:none;mask:none}tui-accordion>[tuiAccordion]:after{margin-inline-start:auto!important}tui-accordion>[tuiAccordion]._open+tui-expand{visibility:visible}tui-accordion>[tuiAccordion]>tui-avatar,tui-accordion>[tuiAccordion]>[tuiAvatar]{margin:0}tui-accordion>tui-expand{visibility:hidden;min-block-size:1px!important;padding:1.25rem;transition-property:grid-template-rows,padding,visibility!important;box-shadow:inset 0 0 0 1px var(--tui-border-normal)}tui-accordion>tui-expand:last-child{border-bottom-left-radius:inherit!important;border-bottom-right-radius:inherit!important}tui-accordion[data-size=m]{font:var(--tui-font-text-s)}tui-accordion[data-size=m]>tui-expand{padding:1rem}tui-accordion[data-size=s]{font:var(--tui-font-text-s)}tui-accordion[data-size=s]>tui-expand{padding:.625rem}\n"] }]
        }], propDecorators: { expands: [{
                type: ContentChildren,
                args: [TuiExpand]
            }], directives: [{
                type: ContentChildren,
                args: [TuiAccordionDirective]
            }], closeOthers: [{
                type: Input
            }], sizeSetter: [{
                type: Input,
                args: ['size']
            }] } });

const TuiAccordion = [
    TuiAccordionComponent,
    TuiAccordionDirective,
    TuiExpand,
    TuiItem,
];

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAccordion, TuiAccordionComponent, TuiAccordionDirective };
