// @taiga-ui/addon-commerce/esm2022/pipes/format-card/format-card.pipe.mjs
import { Pipe } from '@angular/core';
import { CHAR_NO_BREAK_SPACE } from '@taiga-ui/cdk/constants';
import * as i0 from "@angular/core";
class TuiFormatCardPipe {
    transform(value = '', cardPrefilled = false) {
        return value && !cardPrefilled
            ? value
                .split('')
                .map((char, index) => index && index % 4 === 0 ? `${CHAR_NO_BREAK_SPACE}${char}` : char)
                .join('')
            : '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiFormatCardPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiFormatCardPipe, isStandalone: true, name: "tuiFormatCard" }); }
}
export { TuiFormatCardPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiFormatCardPipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiFormatCard',
                }]
        }] });
