// @taiga-ui/addon-doc/esm2022/components/page/page.providers.mjs
import { ElementRef, InjectionToken } from '@angular/core';
import { TUI_DOC_SEE_ALSO } from '@taiga-ui/addon-doc/tokens';
import { tuiCreateOptions } from '@taiga-ui/cdk/utils/di';
export const [TUI_DOC_PAGE_OPTIONS, tuiDocPageOptionsProvider] = tuiCreateOptions({
    header: '',
    package: '',
    type: '',
    tags: [],
    path: '',
    deprecated: false,
});
export const TUI_DOC_TABS = new InjectionToken(ngDevMode ? 'TUI_DOC_TABS' : '', {
    factory: () => () => ({}),
});
/**
 * Array if related page titles
 */
export const PAGE_SEE_ALSO = new InjectionToken(ngDevMode ? 'PAGE_SEE_ALSO' : '');
export const PAGE_PROVIDERS = [
    {
        provide: PAGE_SEE_ALSO,
        deps: [ElementRef, TUI_DOC_SEE_ALSO],
        useFactory: ({ nativeElement }, seeAlsoGroups) => {
            const currentHeader = nativeElement.getAttribute('header');
            const groups = seeAlsoGroups.filter((group) => group.includes(currentHeader)) || [];
            const seeAlsoSet = new Set(groups
                .join()
                .split(',')
                .filter((component) => component && component !== currentHeader));
            return Array.from(seeAlsoSet);
        },
    },
];
