// @taiga-ui/legacy/esm2022/directives/textfield-controller/textfield-label-outside.directive.mjs
import { Directive, InjectionToken, Input } from '@angular/core';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { AbstractTuiController } from '@taiga-ui/legacy/classes';
import * as i0 from "@angular/core";
export const TUI_TEXTFIELD_LABEL_OUTSIDE = new InjectionToken(ngDevMode ? 'TUI_TEXTFIELD_LABEL_OUTSIDE' : '', {
    factory: () => new TuiTextfieldLabelOutsideDirective(),
});
class TuiTextfieldLabelOutsideDirective extends AbstractTuiController {
    constructor() {
        super(...arguments);
        this.labelOutside = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldLabelOutsideDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextfieldLabelOutsideDirective, selector: "[tuiTextfieldLabelOutside]", inputs: { labelOutside: ["tuiTextfieldLabelOutside", "labelOutside"] }, providers: [
            tuiProvide(TUI_TEXTFIELD_LABEL_OUTSIDE, TuiTextfieldLabelOutsideDirective),
        ], usesInheritance: true, ngImport: i0 }); }
}
export { TuiTextfieldLabelOutsideDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldLabelOutsideDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: '[tuiTextfieldLabelOutside]',
                    providers: [
                        tuiProvide(TUI_TEXTFIELD_LABEL_OUTSIDE, TuiTextfieldLabelOutsideDirective),
                    ],
                }]
        }], propDecorators: { labelOutside: [{
                type: Input,
                args: ['tuiTextfieldLabelOutside']
            }] } });
