// @taiga-ui/kit/esm2022/directives/lazy-loading/lazy-loading.directive.mjs
import { isPlatformServer } from '@angular/common';
import { Directive, inject, Input, PLATFORM_ID, signal } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { IntersectionObserverService } from '@ng-web-apis/intersection-observer';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { TuiLazyLoadingService } from './lazy-loading.service';
import * as i0 from "@angular/core";
/**
 * @deprecated: Drop in v5.0
 */
class TuiImgLazyLoading {
    constructor() {
        this.isServer = isPlatformServer(inject(PLATFORM_ID));
        this.loading$ = inject(TuiLazyLoadingService);
        this.supported = 'loading' in tuiInjectElement();
        this.src = signal(null);
        this.background = signal(this.isServer ? '' : 'var(--tui-background-neutral-2)');
        this.animation = signal(this.isServer ? '' : 'tuiSkeletonVibe ease-in-out 1s infinite alternate');
        this.$ = !this.supported &&
            // eslint-disable-next-line @angular-eslint/no-uncalled-signals
            this.loading$.pipe(takeUntilDestroyed()).subscribe((src) => this.src.set(src));
    }
    set srcSetter(src) {
        if (this.supported) {
            this.src.set(src);
        }
        else {
            this.loading$.next(src);
        }
    }
    unset() {
        this.background.set('');
        this.animation.set('');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiImgLazyLoading, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiImgLazyLoading, isStandalone: true, selector: "img[loading=\"lazy\"],img[tuiLoading=\"lazy\"]", inputs: { srcSetter: ["src", "srcSetter"] }, host: { listeners: { "load": "unset()", "error": "unset()" }, properties: { "style.animation": "animation()", "style.background": "background()", "attr.loading": "supported ? \"lazy\" : null", "attr.src": "src()" } }, providers: [TuiLazyLoadingService, IntersectionObserverService], ngImport: i0 }); }
}
export { TuiImgLazyLoading };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiImgLazyLoading, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'img[loading="lazy"],img[tuiLoading="lazy"]',
                    providers: [TuiLazyLoadingService, IntersectionObserverService],
                    host: {
                        '[style.animation]': 'animation()',
                        '[style.background]': 'background()',
                        '[attr.loading]': 'supported ? "lazy" : null',
                        '[attr.src]': 'src()',
                        '(load)': 'unset()',
                        '(error)': 'unset()',
                    },
                }]
        }], propDecorators: { srcSetter: [{
                type: Input,
                args: ['src']
            }] } });
