// @taiga-ui/kit/esm2022/components/toast/toast.service.mjs
import { Directive, inject, Injectable } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { TuiPopoverDirective } from '@taiga-ui/cdk/directives/popover';
import { tuiAsPopover, TuiPopoverService } from '@taiga-ui/cdk/services';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { tuiCreateToken, tuiCreateTokenFromFactory, } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_ALERTS } from '@taiga-ui/core/components/alert';
import { BehaviorSubject, pairwise } from 'rxjs';
import { TuiToastComponent } from './toast.component';
import { TUI_TOAST_OPTIONS } from './toast.options';
import * as i0 from "@angular/core";
const TOASTS = tuiCreateToken(new BehaviorSubject([]));
export const TUI_TOASTS_CONCURRENCY = tuiCreateTokenFromFactory(() => inject(TUI_IS_MOBILE) ? 1 : 2);
class TuiToastService extends TuiPopoverService {
    constructor() {
        super(...arguments);
        this.concurrency = inject(TUI_TOASTS_CONCURRENCY);
        this.alerts = inject(TUI_ALERTS);
        this.sub = this.items$
            .pipe(pairwise(), takeUntilDestroyed())
            .subscribe(([prev, next]) => {
            const closed = prev.filter((item) => !next.includes(item));
            const alerts = this.alerts.value.filter((item) => !closed.includes(item));
            const toasts = new Set([...alerts, ...next.slice(0, this.concurrency)]);
            this.alerts.next(Array.from(toasts));
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiToastService, deps: null, target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiToastService, providedIn: 'root', useFactory: () => new TuiToastService(TOASTS, TuiToastComponent, inject(TUI_TOAST_OPTIONS)) }); }
}
export { TuiToastService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiToastService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                    useFactory: () => new TuiToastService(TOASTS, TuiToastComponent, inject(TUI_TOAST_OPTIONS)),
                }]
        }] });
class TuiToastTemplate extends TuiPopoverDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiToastTemplate, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiToastTemplate, isStandalone: true, selector: "ng-template[tuiToast]", inputs: { options: ["tuiToastOptions", "options"], open: ["tuiToast", "open"] }, outputs: { openChange: "tuiToastChange" }, providers: [tuiAsPopover(TuiToastService)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiToastTemplate };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiToastTemplate, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[tuiToast]',
                    inputs: ['options: tuiToastOptions', 'open: tuiToast'],
                    outputs: ['openChange: tuiToastChange'],
                    providers: [tuiAsPopover(TuiToastService)],
                }]
        }] });
