// @taiga-ui/kit/esm2022/components/progress/progress-bar/progress-color-segments.directive.mjs
import { computed, Directive, inject, Input, signal } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { MutationObserverService, WA_MUTATION_OBSERVER_INIT, } from '@ng-web-apis/mutation-observer';
import { tuiWatch } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { map } from 'rxjs';
import * as i0 from "@angular/core";
class TuiProgressColorSegments {
    constructor() {
        this.colors = signal([]);
        this.el = tuiInjectElement();
        this.position = toSignal(inject(MutationObserverService, { self: true }).pipe(map(() => this.el.position), tuiWatch()), { initialValue: this.el.position });
        this.color = computed(() => {
            const colors = this.colors();
            const position = this.position();
            if (!colors.length || position <= 0) {
                return null;
            }
            const colorsString = colors.reduce((acc, color, i) => `${acc}, ${color} calc(${i} / ${colors.length} * 100% / ${position}) calc(${i + 1} / ${colors.length} * 100% / ${position})`, '');
            return `linear-gradient(to right${colorsString})`;
        });
    }
    set colorsSetter(colors) {
        this.colors.set(colors);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiProgressColorSegments, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiProgressColorSegments, isStandalone: true, selector: "progress[tuiProgressBar][tuiProgressColorSegments]", inputs: { colorsSetter: ["tuiProgressColorSegments", "colorsSetter"] }, host: { properties: { "style.--tui-progress-color": "color()" } }, providers: [
            MutationObserverService,
            {
                provide: WA_MUTATION_OBSERVER_INIT,
                useValue: {
                    attributeOldValue: true,
                },
            },
        ], ngImport: i0 }); }
}
export { TuiProgressColorSegments };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiProgressColorSegments, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'progress[tuiProgressBar][tuiProgressColorSegments]',
                    providers: [
                        MutationObserverService,
                        {
                            provide: WA_MUTATION_OBSERVER_INIT,
                            useValue: {
                                attributeOldValue: true,
                            },
                        },
                    ],
                    host: { '[style.--tui-progress-color]': 'color()' },
                }]
        }], propDecorators: { colorsSetter: [{
                type: Input,
                args: ['tuiProgressColorSegments']
            }] } });
