// @taiga-ui/core/esm2022/components/textfield/select-like.directive.mjs
import { Directive, inject } from '@angular/core';
import { TUI_IS_ANDROID } from '@taiga-ui/cdk/tokens';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import * as i0 from "@angular/core";
class TuiSelectLike {
    constructor() {
        this.el = tuiInjectElement();
        this.isAndroid = inject(TUI_IS_ANDROID);
    }
    clear() {
        this.el.value = '';
    }
    prevent(event) {
        if (!this.isAndroid) {
            return;
        }
        event.preventDefault();
        this.el.focus();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectLike, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSelectLike, isStandalone: true, selector: "[tuiSelectLike]", host: { attributes: { "tuiSelectLike": "", "inputmode": "none", "spellcheck": "false", "autocomplete": "off" }, listeners: { "beforeinput": "$event.inputType.includes(\"delete\") || $event.preventDefault()", "input.capture": "$event.inputType?.includes(\"delete\") && clear()", "mousedown": "prevent($event)" } }, ngImport: i0 }); }
}
export { TuiSelectLike };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectLike, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiSelectLike]',
                    host: {
                        tuiSelectLike: '',
                        inputmode: 'none',
                        spellcheck: 'false',
                        autocomplete: 'off',
                        // Click on cleaner icon does not trigger `beforeinput` event --> handle all kind of deletion in input event
                        '(beforeinput)': '$event.inputType.includes("delete") || $event.preventDefault()',
                        '(input.capture)': '$event.inputType?.includes("delete") && clear()',
                        // Hide Android text select handle (bubble marker below transparent caret)
                        '(mousedown)': 'prevent($event)',
                    },
                }]
        }] });
