// @taiga-ui/legacy/esm2022/components/table-bar/table-bar.options.mjs
import { InjectionToken } from '@angular/core';
import { tuiProvideOptions } from '@taiga-ui/cdk/utils/miscellaneous';
import { BehaviorSubject } from 'rxjs';
/**
 * @deprecated drop in v5.0 use {@link TuiActionBar}
 * https://taiga-ui.dev/components/actions-bar
 */
export const TUI_TABLE_BARS = new InjectionToken(ngDevMode ? 'TUI_TABLE_BARS' : '', {
    factory: () => new BehaviorSubject([]),
});
/**
 * @deprecated drop in v5.0 use {@link TuiActionBar}
 * https://taiga-ui.dev/components/actions-bar
 */
export const TUI_TABLE_BAR_DEFAULT_OPTIONS = {
    adaptive: true,
    hasCloseButton: true,
    mode: 'onDark',
};
/**
 * @deprecated drop in v5.0 use {@link TuiActionBar}
 * https://taiga-ui.dev/components/actions-bar
 */
export const TUI_TABLE_BAR_OPTIONS = new InjectionToken(ngDevMode ? 'TUI_TABLE_BAR_OPTIONS' : '', {
    factory: () => TUI_TABLE_BAR_DEFAULT_OPTIONS,
});
/**
 * @deprecated drop in v5.0 use {@link TuiActionBar}
 * https://taiga-ui.dev/components/actions-bar
 */
export function tuiTableBarOptionsProvider(options) {
    return tuiProvideOptions(TUI_TABLE_BAR_OPTIONS, options, TUI_TABLE_BAR_DEFAULT_OPTIONS);
}
