// @taiga-ui/addon-doc/esm2022/utils/coerce-value.mjs
function isEmptyParamValue(value) {
    return ['NaN', 'null', 'undefined'].includes(value);
}
function isBooleanParamValue(value) {
    return value === 'true' || value === 'false';
}
function isNumberParamValue(value) {
    return (!!String(value).trim() &&
        !Number.isNaN(Number(value)) &&
        !String(value).startsWith('+'));
}
function isPossibleArray(value) {
    return value.startsWith('[') && value.endsWith(']');
}
function isPossibleObject(value) {
    return value.startsWith('{') && value.endsWith('}');
}
export function tuiCoerceValue(value) {
    const prepared = String(value).trim();
    if (isEmptyParamValue(prepared)) {
        return null;
    }
    if (isBooleanParamValue(prepared)) {
        return String(prepared) === 'true';
    }
    if (isNumberParamValue(prepared)) {
        return Number(prepared);
    }
    if (prepared === '%') {
        // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent#exceptions
        return prepared;
    }
    const decodedValue = decodeURIComponent(prepared);
    try {
        return isPossibleArray(decodedValue) || isPossibleObject(decodedValue)
            ? JSON.parse(decodedValue)
            : decodedValue;
    }
    catch {
        return decodedValue;
    }
}
