// @taiga-ui/kit/esm2022/components/input-time/input-time.directive.mjs
import { computed, Directive, inject, Input, signal, untracked } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { MaskitoDirective } from '@maskito/angular';
import { maskitoAddOnFocusPlugin, maskitoCaretGuard, maskitoRemoveOnBlurPlugin, maskitoSelectionChangeHandler, maskitoTimeOptionsGenerator, } from '@maskito/kit';
import { tuiAsControl, TuiControl, tuiValueTransformerFrom } from '@taiga-ui/cdk/classes';
import { TuiTime } from '@taiga-ui/cdk/date-time';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { tuiDirectiveBinding } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiAsOptionContent } from '@taiga-ui/core/components/data-list';
import { tuiAsTextfieldAccessor, TuiTextfieldComponent, TuiTextfieldDirective, tuiTextfieldIconBinding, TuiWithNativePicker, TuiWithTextfield, } from '@taiga-ui/core/components/textfield';
import { TuiDropdownDirective, tuiDropdownEnabled, tuiDropdownOpen, } from '@taiga-ui/core/directives/dropdown';
import { tuiAsAuxiliary } from '@taiga-ui/core/tokens';
import { TuiSelectOption } from '@taiga-ui/kit/components/select';
import { TUI_TIME_TEXTS } from '@taiga-ui/kit/tokens';
import { tuiMaskito } from '@taiga-ui/kit/utils';
import { TUI_INPUT_TIME_OPTIONS } from './input-time.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/components/textfield";
import * as i2 from "@maskito/angular";
class TuiInputTimeDirective extends TuiControl {
    constructor() {
        super(...arguments);
        this.textfield = inject(TuiTextfieldDirective);
        this.dropdown = inject(TuiDropdownDirective);
        this.open = tuiDropdownOpen();
        this.options = inject(TUI_INPUT_TIME_OPTIONS);
        this.fillers = toSignal(inject(TUI_TIME_TEXTS));
        this.prefix = signal('');
        this.postfix = signal('');
        this.icon = tuiTextfieldIconBinding(TUI_INPUT_TIME_OPTIONS);
        this.dropdownEnabled = tuiDropdownEnabled(computed(() => !this.native && this.interactive()));
        this.filler = tuiDirectiveBinding(TuiTextfieldComponent, 'fillerSetter', computed((filler = this.fillers()?.[this.timeMode()] ?? '') => this.postfix() ? '' : this.prefix() + filler), {});
        this.mask = tuiMaskito(computed(() => this.computeMask({
            ...this.options,
            mode: this.timeMode(),
            step: this.interactive() && !this.dropdown._content() ? 1 : 0,
            prefix: this.prefix(),
            postfix: this.postfix(),
        })));
        this.accept = [];
        this.native = !!inject(TuiWithNativePicker, { optional: true }) && inject(TUI_IS_MOBILE);
        this.timeMode = signal(this.options.mode);
    }
    // TODO(v5): use signal inputs
    set modeSetter(x) {
        this.timeMode.set(x);
    }
    // TODO(v5): use signal inputs
    set prefixSetter(x) {
        this.prefix.set(x);
    }
    // TODO(v5): use signal inputs
    set postfixSetter(x) {
        this.postfix.set(x);
    }
    setValue(value) {
        this.onChange(value);
        if (value) {
            this.textfield.value.set(this.stringify(value));
        }
        else {
            this.textfield.setValue(value);
        }
        if (!value && this.dropdownEnabled()) {
            this.open.set(true);
        }
    }
    writeValue(value) {
        const reset = this.control.pristine && this.control.untouched && !value;
        const changed = untracked(() => value !== this.value());
        if (changed || reset) {
            super.writeValue(value);
            untracked(() => this.textfield.value.set(this.stringify(this.value())));
        }
    }
    onInput(valueWithAffixes) {
        const value = valueWithAffixes
            .replace(this.prefix(), '')
            .replace(this.postfix(), '');
        const time = value.length === this.timeMode().length ? TuiTime.fromString(value) : null;
        const newValue = this.accept.length && time ? this.findNearestTime(time, this.accept) : time;
        this.control?.control?.updateValueAndValidity({ emitEvent: false });
        this.onChange(newValue);
        if (newValue && newValue !== time) {
            this.textfield.value.set(this.stringify(newValue));
        }
    }
    toggle() {
        this.open.update((x) => !x);
    }
    computeMask(params) {
        const options = maskitoTimeOptionsGenerator(params);
        const { mode, prefix, postfix } = params;
        const inputModeSwitchPlugin = maskitoSelectionChangeHandler((element) => {
            element.inputMode =
                element.selectionStart >= mode.indexOf(' AA') ? 'text' : 'numeric';
        });
        const caretGuardPlugin = maskitoCaretGuard((value) => [
            prefix.length,
            value.length - postfix.length,
        ]);
        return {
            ...options,
            plugins: options.plugins.concat(caretGuardPlugin, maskitoAddOnFocusPlugin(prefix + postfix), maskitoRemoveOnBlurPlugin(prefix + postfix), mode.includes('AA') ? inputModeSwitchPlugin : []),
        };
    }
    findNearestTime(value, items) {
        // eslint-disable-next-line no-restricted-syntax
        return items.reduce((previous, current) => Math.abs(current.valueOf() - value.valueOf()) <
            Math.abs(previous.valueOf() - value.valueOf())
            ? current
            : previous);
    }
    stringify(time) {
        return this.prefix() + (time?.toString(this.timeMode()) || '') + this.postfix();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTimeDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputTimeDirective, isStandalone: true, selector: "input[tuiInputTime]", inputs: { accept: "accept", modeSetter: ["mode", "modeSetter"], prefixSetter: ["prefix", "prefixSetter"], postfixSetter: ["postfix", "postfixSetter"] }, host: { attributes: { "inputmode": "numeric" }, listeners: { "click": "toggle()", "input": "onInput($event.target.value)" }, properties: { "disabled": "disabled()" } }, providers: [
            tuiAsControl(TuiInputTimeDirective),
            tuiAsTextfieldAccessor(TuiInputTimeDirective),
            tuiAsAuxiliary(TuiInputTimeDirective),
            tuiValueTransformerFrom(TUI_INPUT_TIME_OPTIONS),
            tuiAsOptionContent(TuiSelectOption),
        ], usesInheritance: true, hostDirectives: [{ directive: i1.TuiWithTextfield }, { directive: i2.MaskitoDirective }], ngImport: i0 }); }
}
export { TuiInputTimeDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTimeDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'input[tuiInputTime]',
                    providers: [
                        tuiAsControl(TuiInputTimeDirective),
                        tuiAsTextfieldAccessor(TuiInputTimeDirective),
                        tuiAsAuxiliary(TuiInputTimeDirective),
                        tuiValueTransformerFrom(TUI_INPUT_TIME_OPTIONS),
                        tuiAsOptionContent(TuiSelectOption),
                    ],
                    hostDirectives: [TuiWithTextfield, MaskitoDirective],
                    host: {
                        inputmode: 'numeric',
                        '[disabled]': 'disabled()',
                        '(click)': 'toggle()',
                        '(input)': 'onInput($event.target.value)',
                    },
                }]
        }], propDecorators: { accept: [{
                type: Input
            }], modeSetter: [{
                type: Input,
                args: ['mode']
            }], prefixSetter: [{
                type: Input,
                args: ['prefix']
            }], postfixSetter: [{
                type: Input,
                args: ['postfix']
            }] } });
