// @taiga-ui/experimental/esm2022/components/search-results/search-hotkey.directive.mjs
import { Directive, inject } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { WA_NAVIGATOR } from '@ng-web-apis/common';
import { TuiInputSearch } from '@taiga-ui/layout/components/input-search';
import { TUI_INPUT_SEARCH } from '@taiga-ui/layout/tokens';
import { map } from 'rxjs';
import * as i0 from "@angular/core";
class TuiSearchHotkey {
    constructor() {
        this.platform = inject(WA_NAVIGATOR).platform;
        this.search = inject(TuiInputSearch);
        this.placeholder = toSignal(inject(TUI_INPUT_SEARCH).pipe(map(({ hotkey }) => this.platform.startsWith('Mac') ? `⌘+K ${hotkey}` : `Alt+K ${hotkey}`)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSearchHotkey, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSearchHotkey, isStandalone: true, selector: "input[tuiSearchHotkey]", host: { listeners: { "document:keydown.meta.k.stop": "search.open()", "document:keydown.alt.k.stop": "search.open()" }, properties: { "placeholder": "placeholder()" } }, ngImport: i0 }); }
}
export { TuiSearchHotkey };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSearchHotkey, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'input[tuiSearchHotkey]',
                    host: {
                        '[placeholder]': 'placeholder()',
                        '(document:keydown.meta.k.stop)': 'search.open()',
                        '(document:keydown.alt.k.stop)': 'search.open()',
                    },
                }]
        }] });
