// @taiga-ui/legacy/esm2022/components/color-selector/linear-multi-picker/linear-multi-picker.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLinearMultiPickerComponent } from './linear-multi-picker.component';
import * as i0 from "@angular/core";
/**
 * TODO(v5): delete it
 * @deprecated use new version of {@link https://taiga-ui.dev/components/input-color TuiInputColor} (from @taiga-ui/kit) instead
 */
class TuiLinearMultiPickerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiLinearMultiPickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiLinearMultiPickerModule, declarations: [TuiLinearMultiPickerComponent], imports: [CommonModule], exports: [TuiLinearMultiPickerComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiLinearMultiPickerModule, imports: [CommonModule] }); }
}
export { TuiLinearMultiPickerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiLinearMultiPickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiLinearMultiPickerComponent],
                    exports: [TuiLinearMultiPickerComponent],
                }]
        }] });
