// @taiga-ui/legacy/esm2022/components/input-phone/input-phone.module.mjs
import { NgModule } from '@angular/core';
import { MaskitoDirective } from '@maskito/angular';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiLegacyDropdownOpenMonitorDirective, TuiTextfieldControllerModule, TuiValueAccessorModule, } from '@taiga-ui/legacy/directives';
import { TuiInputPhoneComponent } from './input-phone.component';
import { TuiInputPhoneDirective } from './input-phone.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
/**
 * @deprecated use {@link https://taiga-ui.dev/components/input-phone TuiInputPhone} with {@link https://taiga-ui.dev/components/textfield TuiTextfield}
 */
class TuiInputPhoneModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPhoneModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPhoneModule, declarations: [TuiInputPhoneComponent, TuiInputPhoneDirective], imports: [MaskitoDirective,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiValueAccessorModule,
            TuiLegacyDropdownOpenMonitorDirective, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputPhoneComponent,
            TuiInputPhoneDirective,
            TuiTextfieldComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPhoneModule, imports: [TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiValueAccessorModule, i1.TuiDropdownComponent] }); }
}
export { TuiInputPhoneModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPhoneModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MaskitoDirective,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiValueAccessorModule,
                        TuiLegacyDropdownOpenMonitorDirective,
                        ...TuiDropdown,
                    ],
                    declarations: [TuiInputPhoneComponent, TuiInputPhoneDirective],
                    exports: [
                        TuiInputPhoneComponent,
                        TuiInputPhoneDirective,
                        TuiTextfieldComponent,
                        ...TuiDropdown,
                    ],
                }]
        }] });
