// @taiga-ui/experimental/esm2022/components/accordion/accordion.directive.mjs
import { Directive, EventEmitter, inject, Input, Output, signal, } from '@angular/core';
import { tuiDirectiveBinding } from '@taiga-ui/cdk/utils';
import { TuiButton, tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import { tuiAvatarOptionsProvider } from '@taiga-ui/kit/components/avatar';
import { TuiChevron } from '@taiga-ui/kit/directives/chevron';
import { TuiAccordionComponent } from './accordion.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/components/button";
import * as i2 from "@taiga-ui/kit/directives/chevron";
class TuiAccordionDirective {
    constructor() {
        this.accordion = inject(TuiAccordionComponent);
        this.size = tuiDirectiveBinding(TuiButton, 'size', this.accordion.size);
        this.tuiAccordion = '';
        this.tuiAccordionChange = new EventEmitter();
        this.open = tuiDirectiveBinding(TuiChevron, 'tuiChevron', signal(false));
    }
    ngOnChanges() {
        this.open.set(!!this.tuiAccordion);
        this.accordion.toggle(this);
    }
    toggle() {
        this.open.set(!this.open());
        this.tuiAccordion = this.open();
        this.tuiAccordionChange.emit(this.open());
        this.accordion.toggle(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiAccordionDirective, isStandalone: true, selector: "button[tuiAccordion]", inputs: { tuiAccordion: "tuiAccordion" }, outputs: { tuiAccordionChange: "tuiAccordionChange" }, host: { attributes: { "tuiButton": "", "tuiAccordion": "", "type": "button" }, listeners: { "click": "toggle()" }, properties: { "class._open": "open()" } }, providers: [
            tuiAvatarOptionsProvider({ size: 's' }),
            tuiButtonOptionsProvider({ appearance: 'outline-grayscale' }),
        ], usesOnChanges: true, hostDirectives: [{ directive: i1.TuiButton }, { directive: i2.TuiChevron }], ngImport: i0 }); }
}
export { TuiAccordionDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'button[tuiAccordion]',
                    providers: [
                        tuiAvatarOptionsProvider({ size: 's' }),
                        tuiButtonOptionsProvider({ appearance: 'outline-grayscale' }),
                    ],
                    hostDirectives: [TuiButton, TuiChevron],
                    host: {
                        tuiButton: '',
                        tuiAccordion: '',
                        type: 'button',
                        '[class._open]': 'open()',
                        '(click)': 'toggle()',
                    },
                }]
        }], propDecorators: { tuiAccordion: [{
                type: Input
            }], tuiAccordionChange: [{
                type: Output
            }] } });
