// @taiga-ui/experimental/esm2022/components/accordion/accordion.component.mjs
import { ChangeDetectionStrategy, Component, ContentChildren, DestroyRef, inject, Input, signal, ViewEncapsulation, } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { EMPTY_QUERY } from '@taiga-ui/cdk/constants';
import { TuiGroup, tuiGroupOptionsProvider } from '@taiga-ui/core/directives/group';
import { TuiExpand } from '@taiga-ui/experimental/components/expand';
import { combineLatest, map, merge, ReplaySubject, withLatestFrom } from 'rxjs';
import { TuiAccordionDirective } from './accordion.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/group";
class TuiAccordionComponent {
    constructor() {
        this.destroyRef = inject(DestroyRef);
        this.toggle$ = new ReplaySubject(Infinity);
        this.expands = EMPTY_QUERY;
        this.directives = EMPTY_QUERY;
        this.closeOthers = true;
        this.size = signal('l');
    }
    set sizeSetter(size) {
        this.size.set(size);
    }
    ngAfterViewInit() {
        merge(combineLatest([this.directives.changes, this.expands.changes]).pipe(withLatestFrom(this.toggle$), map(([, accordion]) => accordion)), this.toggle$)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe((accordion) => this.expand(accordion));
    }
    toggle(directive) {
        this.toggle$.next(directive);
    }
    expand(accordion) {
        if (this.closeOthers && accordion.open()) {
            this.expands.forEach((expand) => {
                expand.expanded = false;
            });
            this.directives.forEach((dir) => {
                if (dir === accordion) {
                    return;
                }
                dir.open.set(false);
                dir.tuiAccordion = false;
                dir.tuiAccordionChange.emit(false);
            });
        }
        const expand = this.expands.get(this.directives.toArray().indexOf(accordion));
        if (expand) {
            expand.expanded = accordion.open();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiAccordionComponent, isStandalone: true, selector: "tui-accordion", inputs: { closeOthers: "closeOthers", sizeSetter: ["size", "sizeSetter"] }, host: { properties: { "attr.data-size": "size()" } }, providers: [
            tuiGroupOptionsProvider({
                orientation: 'vertical',
                collapsed: true,
            }),
        ], queries: [{ propertyName: "expands", predicate: TuiExpand }, { propertyName: "directives", predicate: TuiAccordionDirective }], hostDirectives: [{ directive: i1.TuiGroup }], ngImport: i0, template: '<ng-content />', isInline: true, styles: ["tui-accordion{inline-size:-webkit-fill-available;inline-size:-moz-available;inline-size:stretch;border-radius:var(--t-group-radius);font:var(--tui-font-text-m);overflow:hidden}tui-accordion>[tuiAccordion]{text-align:start;justify-content:flex-start}tui-accordion>[tuiAccordion]:last-of-type:not(._open){-webkit-mask-image:var(--t-group-mask-end);mask-image:var(--t-group-mask-end);border-bottom-left-radius:inherit;border-bottom-right-radius:inherit}tui-accordion>[tuiAccordion]:first-child{border-top-left-radius:inherit!important;border-top-right-radius:inherit!important}tui-accordion>[tuiAccordion]:first-child:last-of-type:not(._open){-webkit-mask:none;mask:none}tui-accordion>[tuiAccordion]:after{margin-inline-start:auto!important}tui-accordion>[tuiAccordion]._open+tui-expand{visibility:visible}tui-accordion>[tuiAccordion]>tui-avatar,tui-accordion>[tuiAccordion]>[tuiAvatar]{margin:0}tui-accordion>tui-expand{visibility:hidden;min-block-size:1px!important;padding:1.25rem;transition-property:grid-template-rows,padding,visibility!important;box-shadow:inset 0 0 0 1px var(--tui-border-normal)}tui-accordion>tui-expand:last-child{border-bottom-left-radius:inherit!important;border-bottom-right-radius:inherit!important}tui-accordion[data-size=m]{font:var(--tui-font-text-s)}tui-accordion[data-size=m]>tui-expand{padding:1rem}tui-accordion[data-size=s]{font:var(--tui-font-text-s)}tui-accordion[data-size=s]>tui-expand{padding:.625rem}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
export { TuiAccordionComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'tui-accordion', template: '<ng-content />', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, providers: [
                        tuiGroupOptionsProvider({
                            orientation: 'vertical',
                            collapsed: true,
                        }),
                    ], hostDirectives: [TuiGroup], host: {
                        '[attr.data-size]': 'size()',
                    }, styles: ["tui-accordion{inline-size:-webkit-fill-available;inline-size:-moz-available;inline-size:stretch;border-radius:var(--t-group-radius);font:var(--tui-font-text-m);overflow:hidden}tui-accordion>[tuiAccordion]{text-align:start;justify-content:flex-start}tui-accordion>[tuiAccordion]:last-of-type:not(._open){-webkit-mask-image:var(--t-group-mask-end);mask-image:var(--t-group-mask-end);border-bottom-left-radius:inherit;border-bottom-right-radius:inherit}tui-accordion>[tuiAccordion]:first-child{border-top-left-radius:inherit!important;border-top-right-radius:inherit!important}tui-accordion>[tuiAccordion]:first-child:last-of-type:not(._open){-webkit-mask:none;mask:none}tui-accordion>[tuiAccordion]:after{margin-inline-start:auto!important}tui-accordion>[tuiAccordion]._open+tui-expand{visibility:visible}tui-accordion>[tuiAccordion]>tui-avatar,tui-accordion>[tuiAccordion]>[tuiAvatar]{margin:0}tui-accordion>tui-expand{visibility:hidden;min-block-size:1px!important;padding:1.25rem;transition-property:grid-template-rows,padding,visibility!important;box-shadow:inset 0 0 0 1px var(--tui-border-normal)}tui-accordion>tui-expand:last-child{border-bottom-left-radius:inherit!important;border-bottom-right-radius:inherit!important}tui-accordion[data-size=m]{font:var(--tui-font-text-s)}tui-accordion[data-size=m]>tui-expand{padding:1rem}tui-accordion[data-size=s]{font:var(--tui-font-text-s)}tui-accordion[data-size=s]>tui-expand{padding:.625rem}\n"] }]
        }], propDecorators: { expands: [{
                type: ContentChildren,
                args: [TuiExpand]
            }], directives: [{
                type: ContentChildren,
                args: [TuiAccordionDirective]
            }], closeOthers: [{
                type: Input
            }], sizeSetter: [{
                type: Input,
                args: ['size']
            }] } });
